### PR TITLE
Fix file viewer highlight

### PR DESCRIPTION
**Pull request by wagnerand**
Fix https://github.com/mozilla/addons/issues/288.

### CODE REVIEW

File: static/js/zamboni/files.js [hunk 1]
```diff
@@ -228,19 +228,19 @@ if (typeof SyntaxHighlighter !== 'undefined') {
...
-                            'new null return super switch ' +
-                            'this throw true try typeof var while with ' +
-                            'const let of debugger'
-                            ;
+            var keywords = 'break case catch class const continue debugger' +
+                           'default delete do else enum export extends false finally ' +
+                           'for function if implements import in instanceof ' +
+                           'interface let new null package private protected public' +
+                           'static return super switch this throw true try typeof ' +
+                           'var void while with yield';
```
Comment (wagnerand) [inline static/js/zamboni/files.js]: Updated from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#Keywords

File: static/js/zamboni/files.js [hunk 2]
```diff
@@ -228,19 +228,19 @@ if (typeof SyntaxHighlighter !== 'undefined') {
...
 
             var r = SyntaxHighlighter.regexLib;
 
             this.regexList = [
-                { regex: r.multiLineDoubleQuotedString,                 css: 'string' },            // double quoted strings
-                { regex: r.multiLineSingleQuotedString,                 css: 'string' },            // single quoted strings
+                { regex: /"([^\\"\n]|\\[^])*"/g,                        css: 'string' },            // double quoted strings
+                { regex: /'([^\\'\n]|\\[^])*'/g,                        css: 'string' },            // single quoted strings
```
Comment (wagnerand) [inline static/js/zamboni/files.js]: This restores the old behavior.

File: static/js/zamboni/files.js [hunk 3]
```diff
@@ -228,19 +228,19 @@ if (typeof SyntaxHighlighter !== 'undefined') {
...
+                { regex: /`([^`])*`/g,                                  css: 'string' },            // template literals
```
Comment (wagnerand) [inline static/js/zamboni/files.js]: This recognizes template literals.

File: static/js/lib/syntaxhighlighter/shCore.js [hunk 4]
```diff
@@ -2633,6 +2633,64 @@ sh.Highlighter.prototype = {
...
+    /* Combines functionality of old findMatches(), getMatches() and removeNestedMatches()
+     * functions. Parse a string of code using an array of regular expressions.
+     * http://jmrware.com/articles/2011/sh-bug/SyntaxHighlighterBug.html
+     *
+     * @param {Array} regexList     List of regular expressions.
+     * @param {String} code         Source code.
+     * @return {Array}              Returns list of matches.
+     */
+    findMatchesNew: function(regexList, code) { // Rev:20110216_1700
+        var matches = []; // Return value. Array of Match objects.
```
Comment (wagnerand) [inline static/js/lib/syntaxhighlighter/shCore.js]: This has been taken from http://jmrware.com/articles/2011/sh-bug/SyntaxHighlighterBug.html via https://github.com/SyntaxHighlighter/SyntaxHighlighter/issues/72

File: static/js/lib/syntaxhighlighter/shCore.js [hunk 5]
```diff
@@ -2633,6 +2633,64 @@ sh.Highlighter.prototype = {
...
+                    }
+                } // End if re is object.
+            } // End for loop.
+            if (nextMatch) { // Check if one of the regexes matched?
+                func = nextMatch.r.func ? nextMatch.r.func : defaultAdd;
+                resultMatch = func(nextMatch.m, nextMatch.r); // Allow brush option to customize result.
+                if (typeof(resultMatch) === 'string') // Check if we need a new Match object?
+                    resultMatch = new sh.Match(resultMatch, nextMatch.m.index, nextMatch.r.css); // Yes.
+                if (Array.isArray(resultMatch)) {
+                    matches = matches.concat(resultMatch);
```
Comment (wagnerand) [inline static/js/lib/syntaxhighlighter/shCore.js]: This is a custom fix by me that fixes html inline scripts.

File: static/js/lib/syntaxhighlighter/shCore.js [hunk 6]
```diff
@@ -2633,6 +2633,65 @@ sh.Highlighter.prototype = {
...
+                        }
+                    }
+                } // End if re is object.
+            } // End for loop.
+            if (nextMatch) { // Check if one of the regexes matched?
+                func = nextMatch.r.func ? nextMatch.r.func : defaultAdd;
+                resultMatch = func(nextMatch.m, nextMatch.r); // Allow brush option to customize result.
+                if (typeof(resultMatch) === 'string') // Check if we need a new Match object?
+                    resultMatch = new sh.Match(resultMatch, nextMatch.m.index, nextMatch.r.css); // Yes.
+                // Check if the result is an html inline script, in which case it's wrapped in another array.
```
Comment (wagnerand) [inline static/js/lib/syntaxhighlighter/shCore.js]: This is a custom fix by me that fixes html inline scripts.

**Comment by wagnerand** [comment]
# Before
before
# After
after

**Comment by muffinresearch** [comment]
The author of that function mentions it's an [incomplete solution](https://github.com/syntaxhighlighter/syntaxhighlighter/issues/72#issuecomment-3448964). 

Would it be possible to split out a smaller patch to fix just the comment quote issue,  the function could be added separately but I think that's going to need quite a bit of testing to ensure it doesn't cause other regressions. 

Certainly it's starting to look like the better long term fix is just going to be to entirely replace the highlighting with something that's better maintained - the only nit being the diffing part might make that less easy to do.

**Comment by wagnerand** [comment]
My fix in line 2683 patches that hole.

I tried coming up with a smaller patch, but the problem is really a design flaw when handling nested matches for which their is no easy fix.
Maybe I could patch in the old regexp's, but the syntax highlighting would still be completely off for template literals.

**Comment by muffinresearch** [comment]
OK. So I suppose the next step really is to land this and test it thoroughly on -dev to make sure it doesn't introduce more problems with all the various filetypes that review tools needs to handle.

Given there's more to this patch than a couple of lines I'd prefer to wait until the Dec 1st for pushing this out - do you think reviewers can live with the current highlighting issues until then?

File: static/js/zamboni/files.js [hunk 7]
```diff
@@ -228,19 +228,19 @@ if (typeof SyntaxHighlighter !== 'undefined') {
...
+                           'interface let new null package private protected public' +
+                           'static return super switch this throw true try typeof ' +
+                           'var void while with yield';
 
             var r = SyntaxHighlighter.regexLib;
 
             this.regexList = [
                 { regex: r.multiLineDoubleQuotedString,                 css: 'string' },            // double quoted strings
                 { regex: r.multiLineSingleQuotedString,                 css: 'string' },            // single quoted strings
+                { regex: /`([^`])*`/g,                                  css: 'string' },            // template literals
```
Comment (muffinresearch) [inline static/js/zamboni/files.js]: Does this need a capture group?

**Comment by muffinresearch** [comment]
r+wc

**Comment by wagnerand** [comment]
I think pushing this on Dec 1 is ok. I already tested with some files locally and yes, more testing would be appreciated, especially code that uses comments and quote characters intertwined.

**Pull request merged**